refactor(package-manager)!: extend the API and provide access to spawn child

byCedric opened the pull request:
# Why

Fixes ENG-7545

This proposal PR changes the API of the (node) package managers, with the goal of:
  - Having the right API options to install/remove packages to `dependencies`/`devDependencies`/global deps
  - Having access to the actual `spawnAsync` child, to pipe logs into different systems (for example in EAS build)
  - Allowing us to fix a couple of annoying DX issues (see follow up PRs)

> This is further outlined in the Package Manager API proposal.

Before merging this one, we need to merge these other PRs to not-break `main`. Any feedback on the proposal is welcome, especially while in draft.
- https://github.com/expo/expo/pull/19343
    - https://github.com/expo/expo/pull/19344

# How

See Package Manager API proposal.

# Test Plan

> These changes are backward incompatible and require a new version update.

This change also allowed us to implement the following fixes (all stacked PRs):
- https://github.com/expo/expo/pull/19343
    - https://github.com/expo/expo/pull/19344
    - https://github.com/expo/expo/pull/19342
    - https://github.com/expo/expo/pull/19340
    - https://github.com/expo/expo/pull/19341
- https://github.com/expo/expo/pull/19764
- https://github.com/expo/expo/pull/19835
- https://github.com/expo/expo/pull/21189

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

byCedric committed:
refactor!: implement new proposed package manager api

byCedric committed:
[package-manager] Implement pending promise for child access

This pending promise is equal to a delayed spawn async response. We need the `.child` property to be available when executing a method, but still be able to run an async task before that. Making the `.child` property a promise as well.

byCedric committed:
refactor: finalize the proposal implementation

byCedric committed:
refactor: swap out old files for proposal

byCedric committed:
fix: make cocoapods tests work on windows

byCedric committed:
refactor: silence log results for package manager

byCedric committed:
refactor: swap out hacky env for mocking in cocoa test

byCedric committed:
refacrtor: use backwards compatible name for initialization

byCedric committed:
docs: update usage for package manager

byCedric committed:
chore: update build files

byCedric committed:
chore: finalize build and pr

byCedric committed:
docs(package-manager): add missing docs to log and silent options

EvanBacon reviewed (approved):
createForProject being sync is awesome

byCedric commented:
> createForProject being sync is awesome

Yeah, that should be way easier to use. That, and the `manager.workspaceRoot` being sync as well. Whenever we need to interact with that, we can just use stuff like:

```js
await (manager.workspaceRoot() ?? manager).installAsync()
```

byCedric commented on packages/@expo/package-manager/src/node/BasePackageManager.ts:
```diff
@@ -0,0 +1,91 @@
+    const cwd = this.ensureCwdDefined('removeLockFile');
+    const filePath = path.join(cwd, this.lockFile);
+    if (fs.existsSync(filePath)) {
+      rimraf.sync(filePath);
```
Maybe we should drop `rimraf` in favor of `fs`, but for now I left it in here.

byCedric commented on yarn.lock:
```diff
@@ -3086,6 +3086,26 @@
   dependencies:
     "@octokit/openapi-types" "^12.10.0"
 
+"@pnpm/constants@6.1.0":
+  version "6.1.0"
+  resolved "https://registry.yarnpkg.com/@pnpm/constants/-/constants-6.1.0.tgz#2db43ae0e029095df7959bc640081beae38a631b"
+  integrity sha512-L6AiU3OXv9kjKGTJN9j8n1TeJGDcLX9atQlZvAkthlvbXjvKc5SKNWESc/eXhr5nEfuMWhQhiKHDJCpYejmeCQ==
+
+"@pnpm/error@3.0.1":
+  version "3.0.1"
+  resolved "https://registry.yarnpkg.com/@pnpm/error/-/error-3.0.1.tgz#dc232f1717eba895eb842624755c753599b10c9b"
+  integrity sha512-hMlbWbFcfcfolNfSjKjpeaZFow71kNg438LZ8rAd01swiVIYRUf/sRv8gGySru6AijYfz5UqslpIJRDbYBkgQA==
+  dependencies:
+    "@pnpm/constants" "6.1.0"
+
+"@pnpm/find-workspace-dir@^4.0.2":
+  version "4.0.2"
+  resolved "https://registry.yarnpkg.com/@pnpm/find-workspace-dir/-/find-workspace-dir-4.0.2.tgz#347130cb8dead805e6593404c4eae946bcafa8a9"
+  integrity sha512-gU7ycFSWuEGJh7RE/STa33Ch27geODTXIfc+ntiE1BietxfpJIAk34zz51kTUuCFthBkpHlO6yV7jgHD2Tuc3g==
+  dependencies:
+    "@pnpm/error" "3.0.1"
+    find-up "^5.0.0"
+
```
I need to drop this before merging. They aren't required because these methods are `async` methods. We use the same, but a `sync` version of it.

byCedric committed:
refactor: drop pnpm packages from lockfile

byCedric committed:
chore(package-manager): update built files

byCedric committed:
chore(package-manager): fix cocoapod test linting issues

byCedric committed:
refactor(cli): implement new package manager (#19343)

* refactor(cli): implement new package manager

* fix(cli): update ngrok to use log instead of logger

byCedric committed:
fix(package-manager): make error output visible when not silent (#19340)

* refactor(cli): implement new package manager

* fix(cli): update ngrok to use log instead of logger

* fix(package-manager): make error output visible when not running with silent

* test(package-manager): fix old syntax in tests

byCedric committed:
fix(package-manager): disable pnpm frozen lockfiles in CI for prebuild (#19341)

* refactor(cli): implement new package manager

* fix(cli): update ngrok to use log instead of logger

* fix(package-manager): disable pnpm frozen lockfiles in CI for prebuild

byCedric commented:
How do we do the versioning for this package manager change? @EvanBacon It should be a breaking change, luckily, it used the `0.0.x` version, and that technically means every change is a breaking change. But just to be sure, might be better to use `0.1.x` for this change.

byCedric committed:
fix(cli): install dev dependencies when fixing packages (#19344)

* refactor(cli): implement new package manager

* fix(cli): update ngrok to use log instead of logger

* fix(cli): install dev dependencies when fixing packages

* test(cli): add a test for groupBy

byCedric committed:
fix(package-manager): validate if pnpm workspace is part of monorepo (#19342)

* refactor(cli): implement new package manager

* fix(cli): update ngrok to use log instead of logger

* fix(package-manager): validate if pnpm workspace is part of monorepo

* chore(package-manager): update built files

byCedric committed:
refactor(package-manager): rename configAsync to getConfigAsync

This allows us to add `setConfigAsync` in the future.

byCedric committed:
chore(package-manager): rebuild files

byCedric committed:
docs(package-manager): add changelog

byCedric committed:
docs(cli): add missing changelog items

byCedric committed:
test(package-manager): simplify mocking volumes

byCedric committed:
refactor(package-manager): simplify CI detection for pnpm

byCedric committed:
test(package-manager): always disable CI for all pnpm tests

byCedric committed:
refactor(package-manager): replace rimraf with node rm (#19764)

# Why

See [this
comment](https://github.com/expo/expo-cli/pull/4346#discussion_r870063282)

# How

Removed `rimraf` and `fs.existsSync` checks, replaced with
`fs.promises.rm`

> From the docs:
> <img width="1405" alt="image"
src="https://user-images.githubusercontent.com/1203991/198645308-1ec60265-3534-44cd-836a-4b8e831fb10e.png">

- `force` allows us to remove the `existsSync` check. If it doesn't
exist, it just continues without throwing.
- `recursive` is required for non-empty folders, for `node_modules`.

# Test Plan

See tests:
-
[`npm.uninstallAsync`](https://github.com/expo/expo/blob/%40bycedric/package-manager/drop-rimraf/packages/%40expo/package-manager/src/node/__tests__/NpmPackageManager-test.ts#L137-L167)
-
[`pnpm.uninstallAsync`](https://github.com/expo/expo/blob/%40bycedric/package-manager/drop-rimraf/packages/%40expo/package-manager/src/node/__tests__/PnpmPackageManager-test.ts#L174-L204)
-
[`yarn.uninstallAsync`](https://github.com/expo/expo/blob/%40bycedric/package-manager/drop-rimraf/packages/%40expo/package-manager/src/node/__tests__/YarnPackageManager-test.ts#L133-L163)

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg:
https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg:
updated a module plugin).

byCedric committed:
fix(package-manager): add missing default environment variables (#19835)

# Why

~~When `NODE_ENV=production`, pnpm skips installing dev dependencies.
This default isn't right for React Native in general, so we need to
compensate for that.~~

This seems to be a collective behavior, we probably should warn against
this in expo-doctor instead? It doesn't make sense to set `NODE_ENV` to
production for React Native apps.

- [npm docs](https://docs.npmjs.com/cli/v9/commands/npm-install):
> With the `--production` flag (or when the `NODE_ENV` environment
variable is set to `production`), npm will not install modules listed in
`devDependencies`. To install all modules listed in both `dependencies`
and `devDependencies` when `NODE_ENV` environment variable is set to
`production`, you can use `--production=false`.

- [pnpm docs](https://pnpm.io/cli/install#--prod--p):
> pnpm will not install any package listed in `devDependencies` if the
`NODE_ENV` environment variable is set to production. Use this flag to
instruct pnpm to ignore `NODE_ENV` and take its production status from
this flag instead.

- [yarn
docs](https://classic.yarnpkg.com/lang/en/docs/cli/install/#toc-yarn-install-production-true-false):
> Yarn will not install any package listed in `devDependencies` if the
`NODE_ENV` environment variable is set to `production`. Use this flag to
instruct Yarn to ignore `NODE_ENV` and take its production-or-not status
from this flag instead.

# How

- Added `getDefaultEnvironment()` method to each package manager

# Test Plan

- See tests

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg:
https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg:
updated a module plugin).

byCedric committed:
fix(package-manager): install dist-tag versioned dependencies as unversioned (#21189)

# Why

Fixes ENG-7533

# How

We have a workaround in the npm package manager that installs exact
versions of packages differently. This was added to avoid an odd
behavior with npm where the exact version might be skipped, when it
exists as deeply nested dependency.

The workaround writes the exact version directly to `package.json`, and
just runs `npm install`.

While this worked great as workaround, it should not do so when using
`expo@next`. This ignores dist-tag referred dependencies and uses `npm
install --save(-dev)` instead.

# Test Plan

- `$ npm create expo-app ./test`
- `$ cd test`
- `$ npx expo install expo@next` should install the proper dist-tag
version (not `next` as version in `package.json`).


# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg:
https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg:
updated a module plugin).

byCedric commented on packages/@expo/package-manager/CHANGELOG.md:
```diff
@@ -4,12 +4,21 @@
+- Validate if pnpm workspace is part of monorepo. ([#19342](https://github.com/expo/expo/pull/19342) by [@byCedric](https://github.com/byCedric))
+- Present error output to users when not using silent mode. ([#19340](https://github.com/expo/expo/pull/19340) by [@byCedric](https://github.com/byCedric))
+- Disable pnpm frozen lockfiles in CI for prebuild. ([#19341](https://github.com/expo/expo/pull/19341) by [@byCedric](https://github.com/byCedric))
+- Install dist-tag referred dependencies normally as unversioned dependency ([#21189][https://github.com/expo/expo/pull/21189] by [@byCedric](https://github.com/byCedric))
```
im going to hide this, it contains multiple PRs.
